File: python/meetup/meetup.py
from calendar import Calendar

# subclassing the built-in ValueError to create MeetupDayException
class MeetupDayException(ValueError):
    """Exception raised when the Meetup weekday and count do not result in a valid date.

    message: explanation of the error.

    """
    def __init__(self, message):
        self.message = message


def meetup(year, month, week, day_of_week):
    cal = Calendar()
    weekdays = {"Monday": 0,
                "Tuesday": 1,
                "Wednesday": 2,
                "Thursday": 3,
                "Friday": 4,
                "Saturday": 5,
                "Sunday": 6, }

    week_specifiers = {"first": 0,
                       "second": 1,
                       "third": 2,
                       "fourth": 3,
                       "fifth": 4,
                       "last": -1,}

    valid_dates = [d for d in cal.itermonthdates(year, month) if
                   d.month == month and
                   d.weekday() == weekdays[day_of_week]]

    if week in week_specifiers:
        if len(valid_dates) > week_specifiers[week]:
            return valid_dates[week_specifiers[week]]
        else:
            raise MeetupDayException("That day does not exist.")

    if week == "teenth":
        teenth = [d for d in valid_dates if 13 <= d.day <= 19]
        assert len(teenth) == 1
        return teenth[0]

    raise MeetupDayException("That day does not exist.")

File: csharp/meetup/Meetup.cs
using System;
using System.Linq;

public enum Schedule
{
    Teenth,
    First,
    Second,
    Third,
    Fourth,
    Last
}

internal record CurrMonth
{
    public CurrMonth(int year, int month, int day)
    {
        dt = new DateTime(year, month, day);
        dayOfMonth = day;
        dayOfWeek = dt.DayOfWeek;
    }
    
    public DateTime dt { get; }
    public int dayOfMonth { get; }
    public DayOfWeek dayOfWeek { get; }
}

public class Meetup
{
    CurrMonth[] dates { get; }
    
    public Meetup(int month, int year)
    {
        var daysThisMonth = DateTime.DaysInMonth(year, month);
        var days = Enumerable.Range(1, daysThisMonth);
        dates = days.Select(d => new CurrMonth(year, month, d)).ToArray();
    }

    public DateTime Day(DayOfWeek dayOfWeek, Schedule schedule)
    {
        var validDates = dates.Where(d => d.dayOfWeek == dayOfWeek);
        return schedule switch
        {
            Schedule.Teenth => validDates.First(d => d.dayOfMonth is >= 13 and <= 19).dt,
            Schedule.First => validDates.First().dt,
            Schedule.Second => validDates.Skip(1).First().dt,
            Schedule.Third => validDates.Skip(2).First().dt,
            Schedule.Fourth => validDates.Skip(3).First().dt,
            Schedule.Last => validDates.Last().dt,
            _ => throw new ArgumentException("Invalid schedule")
        };
    }
}